### ISSUE TITLE
Compound component (e.g., `Form.Group`, `SearchField.Advanced`) exports include entire module in Webpack bundles

### ISSUE THREAD

**Issue by ochavarria2u** [openedx/paragon#2425]
Hello team, making some investigation I found that some of your imports bundles the entire lib, and the tree shacking is not working. Here's an example of search vs container. 

|Search field: 

- ![image](https://github.com/openedx/paragon/assets/131798747/8e8a81e5-4721-4e45-820f-7135a61f2753)

- ![image](https://github.com/openedx/paragon/assets/131798747/7d854fda-16b5-47f6-af4e-374856b5288b)

Container: 

- ![image](https://github.com/openedx/paragon/assets/131798747/965bad2f-f8b8-492b-a2ce-b4a90ae33fe3)

- ![image](https://github.com/openedx/paragon/assets/131798747/796c5e23-eb19-4ded-9882-afb276cb8b55)

I already validated this with @adamstankiewicz, so please let me know if there's anything I can do in order to help with this issue.

**Comment by adamstankiewicz** [comment]
Thanks for the report of this issue, @ochavarria2u!

We'll need to decide what we want to do to mitigate these bundle size impacts across Paragon's consuming repos in a reasonable way. For example, is there any workaround to keep compound components (`Form.Group`) still functional with tree shaking? Do we have standalone components exported for all compound components that be migrated over to? Is there a CLI tool we could expose to consumers to auto-migrate compound components to their associated standalone components (e.g., `Form.Group` -> `FormGroup`)?

**Comment by viktorrusakov** [comment]
Hey, @ochavarria2u, thanks for reporting this!

The root problem was that our icons package did not support tree-shaking at all, which resulted in all of the icons being included into the initial bundle if your code used at least one of the icons. That's what happened in your first example - `SearchField` uses couple of icons internally which led to including all of the icons into the initial bundle. `Container` on the other hand does not use any icons, so only this component got into the bundle.

Note however that this only caused size problems for the initial bundle size (`Stat size` on the screenshots), but after minification process all of the redundant icons were removed and did not get into the final bundle provided you built your project with the help of `frontend-build` (you can see it by the `Parsed size` and `Gzipped size`). You can read more on how that happened in the comments to the [PR that adds tree-shaking to icons package](https://github.com/openedx/paragon/pull/2441). The fix was released in `v20.45.2` of Paragon and it should drastically reduce `Stat size` of you bundles.

I'm not sure why your first example reports 0B for parsed and gzipped sizes though, I believe that means that none of paragon code got into the final bundle, which is really strange. When I tried to reproduce your example this is what I got
<img width="384" alt="image" src="https://github.com/openedx/paragon/assets/52399399/e216c5a6-2644-4ffc-95ff-e6d6f2892f35">

Could you maybe provide a bit more context on the first example, e.g. how did you use `TestParagonLib` component in the code and what script you used to build the application?

**Comment by ochavarria2u** [comment]
Hi @viktorrusakov thanks for the fix, I've been fetching the reports and it seems that the bundle size got reduced by 150mb. Checking our reports it seems that the parsed size and gzipped size is still 0. I belief this is the reason [why](https://github.com/webpack-contrib/webpack-bundle-analyzer#i-dont-see-gzip-or-parsed-sizes-it-only-shows-stat-size). 
<img width="1748" alt="image" src="https://github.com/openedx/paragon/assets/131798747/d8186934-6aba-4f31-aa18-b7889a87fdd8">

I'm seeing that some of the parents aren't supporting treeshake so we can only do compound components, do you have plans on doing it ? Here's some componenets I've identify that could use the change. 

import { SearchField } from '@edx/paragon';
import { Form } from '@edx/paragon';
import { Collapsible } from '@edx/paragon'; // Needs to implement treeshake
import { Dropdown } from '@edx/paragon'; // Needs to implement treeshake
import { ModalDialog } from '@edx/paragon'; // Needs to implement treeshake
import { Card } from '@edx/paragon'; // Needs to implement treeshake 

FYI - Here's how we use the reports: 
`    ...(process.env.BUNDLE_REPORT ? [
      {
        resolve: 'gatsby-plugin-webpack-bundle-analyser-v2',
        options: {
          devMode: true,
          analyzerMode: 'static',
          openAnalyzer: true,
        },
      },
      'gatsby-plugin-perf-budgets',
    ] : []),`

**Comment by viktorrusakov** [comment]
@ochavarria2u I don't think these component would affect the bundle very much as they only include related compound components (which shouldn't take too much space) into the bundle and not the whole Paragon. However, I understand that's it is still an issue and we do plan to fix them.

I've created a separate issue #2464 to track progress on migrating off of compound components in Paragon. I'm pretty sure there is no way to support tree-shaking for them, so we just have to remove them and instead only use standalone components. I'm not sure when this will be done though, I'd like @adamstankiewicz input here on prioritization.

**Comment by adamstankiewicz** [comment]
@ochavarria2u @viktorrusakov I agree, it appears that compound components (e.g., `Form.Group`, `SearchField.Advanced`) can not support tree-shaking, unfortunately. Thanks Viktor for filing #2464.

I think the first step here should be to evaluate whether all Paragon's compound components have an associated standalone component export (e.g., If we have `Form.Group`, `FormGroup` should also be exported. Assuming we do export all compound components as standalone components today, @ochavarria2u you should be able to migrate off the compound components in your repo to mitigate the tree shaking concern. Though as Viktor mentioned, these components shouldn't really take up too, too much space although we should still definitely optimize it.

Then, tackling what's proposed in #2464 makes sense. We should definitely deprecate the compound components (perhaps even with a legit DEPR ticket, which I'd be happy to file) first with adequate warning for consumers. @viktorrusakov I'll add some thoughts to this effect to #2464.

tl;dr; For the purpose of this issue, we should confirm that all compound components do indeed have an associated standalone components exported already to ensure consumers have a path forward today to mitigate these concerns, and perhaps update the Paragon README about tree shaking to mention lack of support for it with compound components: https://github.com/openedx/paragon#treeshaking

**Comment by brian-smith-tcril** [comment]
Moving conversation to #2464
